chasm/tree.go

```diff
@@ -137,6 +152,378 @@ func NewEmptyTree(
+			}
+			*removedPaths = append(*removedPaths, path)
+			// If parent is about to be removed, it must not have any children.
+			// TODO: softassert: len(childNode.children)==0
+			delete(n.children, childName)
+		}
+	}
+	return nil
+}
+
+func (n *Node) serializeDataNode() error {
+	protoValue, ok := n.value.(proto.Message)
+	if !ok {
+		return serviceerror.NewInternal("only support proto.Message as chasm data")
+	}
+
+	blob, err := serialization.ProtoEncodeBlob(protoValue, enumspb.ENCODING_TYPE_PROTO3)
+	if err != nil {
+		return err
+	}
+
+	n.serializedValue.GetDataAttributes().Data = blob
+	return nil
+}
+
+// deserialize initializes the node's value from its serializedValue.
+// If value is of component type, it initializes every chasm.Field of it and sets node field but not value field.
+// deserialize is not recursive; i.e., must be called on every node separately.
+func (n *Node) deserialize(
+	valueT reflect.Type,
+) error {
+	if err := validateType(valueT); err != nil {
+		return err
+	}
+
+	switch n.serializedValue.Attributes.(type) {
+	case *persistencespb.ChasmNode_ComponentAttributes:
+		return n.deserializeComponentNode(valueT)
+	case *persistencespb.ChasmNode_DataAttributes:
+		return n.deserializeDataNode(valueT)
+	case *persistencespb.ChasmNode_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNode_PointerAttributes:
+		// TODO: return serviceerror.NewInternal(...) instead.
+	}
+	return nil
+}
+
+func (n *Node) deserializeComponentNode(
+	valueT reflect.Type,
+) error {
+	valueV := reflect.New(valueT.Elem())
+	protoMessageFound := false
+	for i := 0; i < valueT.Elem().NumField(); i++ {
+		fieldV := valueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+		if fieldT.AssignableTo(protoMessageT) {
+			if protoMessageFound {
+				return serviceerror.NewInternal("only one proto field allowed in component")
+			}
+			protoMessageFound = true
+
+			value, err := n.unmarshalProto(n.serializedValue.GetComponentAttributes().GetData(), fieldT)
+			if err != nil {
+				return err
+			}
+			fieldV.Set(value)
+			continue
+		}
+
+		// chasm.Field field must be a pointer, i.e. *chasm.Field[T].
+		// TODO: for collection it is not true though.
+		if fieldT.Kind() != reflect.Ptr {
+			continue
+		}
+
+		fieldN := fieldName(valueT.Elem().Field(i))
+
+		switch genericTypePrefix(fieldT.Elem()) {
+		case chasmFieldTypePrefix:
+			if childNode, found := n.children[fieldN]; found {
+				chasmFieldV := reflect.New(fieldT.Elem()) // TODO: support chasm.Field[interface]
+				internalValue := reflect.ValueOf(fieldInternal{
+					node: childNode,
+				})
+				chasmFieldV.Elem().FieldByName(internalFieldName).Set(internalValue)
+				fieldV.Set(chasmFieldV)
+			}
+
+		case chasmCollectionTypePrefix:
+			// TODO: support collection
+			// init the map and populate
+			panic("not implemented")
+		}
+
+		// TODO: error or just go around?
+		// return serviceerror.NewInternal("unsupported field type in component " + fieldT.String())
```

Comment:
let's be strict in the beginning I'd say. It's always easier to relax a check than enforcing one.